src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -611,11 +610,30 @@ public T WithAdfsAuthority(string authorityUri, bool validateAuthority = true)
         public T WithB2CAuthority(string authorityUri)
         {
             var authorityInfo = AuthorityInfo.FromB2CAuthority(authorityUri);
-            Config.Authority = B2CAuthority.CreateAuthority(authorityInfo);
+            Config.Authority = Authority.CreateAuthority(authorityInfo);
 
             return this as T;
         }
 
+        /// <summary>
+        /// Adds a known authority corresponding to a generic OpenIdConnect Identity Provider. 
+        /// MSAL will append ".well-known/openid-configuration" to the authority and retrieve the OIDC 
+        /// metadata from there, to figure out the endpoints.
+        /// See https://openid.net/specs/openid-connect-core-1_0.html#Terminology
+        /// </summary>
+        /// <param name="authorityUri">OpenIdConnect authority</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        /// <remarks>This is an experimental API and only AcquireTokenForClient (client_credentials flow) has been implemented</remarks>        
+        public T WithGenericAuthority(string authorityUri)
```

Comment:
This is placed in the abstract app builder and it is available to the PublicClientApplicationBuilder right? Maybe it is better to put this in ConfidentialClientApplicationBuilder to avoid exposing this to public clients since they dont implement the client_credentials flow

Reply:
Is there a plan to move this to public clients?

Reply:
@trwalke I did some tests to use the GenericAuthority with a public client using an authorization_code flow and minor changes were required. 
@bgavrilMS would you be open to a follow up PR that enable GenericAuthority support to public clients ?

Reply:
@g7ed6e - absolutely! Separate PR please, this one is already huge.